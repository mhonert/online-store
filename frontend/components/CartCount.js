import React from 'react';
import styled from 'styled-components';
import { TransitionGroup, CSSTransition } from 'react-transition-group';

const AnimationStyles = styled.span`
  position: relative;
  .count {
    display: block;
    position: relative;
    transition: all 0.4s;
    backface-visibility: hidden;
  }
  
  .count-enter {
    transform: scale(0.5) rotateX(0.5turn);
  }
  
  .count-enter-active {
    transform: rotateX(0)
  }
  
  .count-exit {
    top: 0;
    position: absolute;
    transform: rotateX(0);
  }
  
  .count-exit-active {
    transform: scale(0.5) rotateX(0.5turn)
  }
  
`;

const Dot = styled.div`
  background: ${props => props.theme.primary};
  color: white;
  border-radius: 50%;
  padding: 0.5rem;
  line-height: 2rem;
  min-width: 3rem;
  margin-left: 1rem;
  font-weight: 100;
  font-feature-settings: 'tnum';
  font-variant-numeric: tabular-nums;
`;

const CartCount = ({ cart }) => {
  const count = countItems(cart);
  return (
    <AnimationStyles>
      <TransitionGroup>
        <CSSTransition
          unmountOnExit
          className="count"
          classNames="count"
          key={count}
          timeout={{ enter: 400, exit: 400 }}
        >
          <Dot>{count}</Dot>
        </CSSTransition>
      </TransitionGroup>
    </AnimationStyles>
  );
};

export const countItems = cart =>
  cart.reduce((count, cartItem) => count + cartItem.quantity, 0);

export default CartCount;
