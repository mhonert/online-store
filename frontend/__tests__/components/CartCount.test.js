import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';
import CartCount from '../../components/CartCount';

const testCart = [{quantity: 1}, {quantity: 2}];

describe('<CartCount/>', () => {
  it('renders', () => {
    shallow(<CartCount cart={testCart} />);
  });

  it('matches the snapshot', () => {
    const wrapper = shallow(<CartCount cart={testCart} />);
    expect(toJSON(wrapper)).toMatchSnapshot();
  });
});