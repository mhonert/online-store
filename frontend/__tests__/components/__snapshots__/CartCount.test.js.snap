// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CartCount/> matches the snapshot 1`] = `
<CartCount__AnimationStyles>
  <TransitionGroup
    childFactory={[Function]}
    component="div"
  >
    <CSSTransition
      className="count"
      classNames="count"
      key="3"
      timeout={
        Object {
          "enter": 400,
          "exit": 400,
        }
      }
      unmountOnExit={true}
    >
      <CartCount__Dot>
        3
      </CartCount__Dot>
    </CSSTransition>
  </TransitionGroup>
</CartCount__AnimationStyles>
`;
